### ISSUE TITLE
false positive E1137 with optional list

### ISSUE THREAD

**Issue by ZeeD** [PyCQA/pylint#3832]
I have this little module:

```python
'xxx'

import typing

def main() -> None:
    '...'

    rows: typing.Iterable[typing.List[str]] = [['foo', 'bar'], ['', 'baz']]

    itor = iter(rows)
    prev: typing.Optional[typing.List[str]] = None
    while True:
        try:
            row = next(itor)
        except StopIteration:
            break

        if row[0]:
            if prev is not None:
                print(prev)
            prev = row
        else:
            assert prev is not None

            for i, cell in enumerate(row):
                if i == 0:
                    continue

                prev[i] += ' ' + cell

    if prev is not None:
        print(prev)

if __name__ == '__main__':
    main()
```

while it does what I expect (it prints `['foo', 'bar baz']`), pylint says that there is an error:

```bat
>pylint falsepositive.py
************* Module falsepositive
falsepositive.py:29:16: E1137: 'prev' does not support item assignment (unsupported-assignment-operation)

------------------------------------------------------------------
Your code has been rated at 7.83/10 (previous run: 7.83/10, +0.00)
```

### pylint --version output
```bat
>pylint --version
pylint 2.6.0
astroid 2.4.2
Python 3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 23:09:28) [MSC v.1916 64 bit (AMD64)]
```

**Comment by clavedeluna** [comment]
I looked into this and the reason is decidedly because `prev`, assigned as `prev: typing.Optional[typing.List[str]] = None` is initialized as None. While `unsupported-assignment-operation` is a red herring, it's telling the author to not initialize a value intended to be a list to a `None` type, instead initialize to `[]`.

Would be nice if pylint had (does it have??) instead a warning to let us know if we've initialized the incorrect empty value for the type. But I'm not sure the particular issue as posed here is worth fixing?
